src/Raven.Client/Documents/Operations/Attachments/GetAttachmentsOperation.cs

```diff
@@ -0,0 +1,141 @@
+                                writer.WriteComma();
+                                writer.WritePropertyName("Value");
+                                writer.WriteString(kvp.Value);
+                                writer.WriteEndObject();;
+
+                            }
+                            writer.WriteEndArray();
+
+                            writer.WriteEndObject();
+                        }
+                    })
+                };
+
+                return request;
+            }
+
+            public override async Task<ResponseDisposeHandling> ProcessResponse(JsonOperationContext context, HttpCache cache, HttpResponseMessage response, string url)
+            {
+                AttachmentsDetails attachmentsMetadata;
+                AttachmentsStreamInfo streamInfo;
+                var state = new JsonParserState();
+                Stream stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
+
+                using (context.GetManagedBuffer(out JsonOperationContext.ManagedPinnedBuffer buffer))
+                using (var parser = new UnmanagedJsonParser(context, state, "attachments/receive"))
+                using (var builder = new BlittableJsonDocumentBuilder(context, BlittableJsonDocumentBuilder.UsageMode.None, "attachments/list", parser, state))
+                using (var peepingTomStream = new PeepingTomStream(stream, context))
+                {
+                    if (await UnmanagedJsonParserHelper.ReadAsync(peepingTomStream, parser, state, buffer).ConfigureAwait(false) == false)
+                        throw new Exception("cannot parse stream");
+
+                    if (state.CurrentTokenType != JsonParserToken.StartObject)
+                        throw new Exception($"Expected token {nameof(JsonParserToken.StartObject)}, but got {nameof(state.CurrentTokenType)}.");
+
+                    await UnmanagedJsonParserHelper.ReadObjectAsync(builder, peepingTomStream, parser, buffer).ConfigureAwait(false);
+
+                    BlittableJsonReaderObject data = builder.CreateReader();
+                    attachmentsMetadata = JsonDeserializationClient.AttachmentAdvancedDetails(data);
+
+                    var bufferSize = parser.BufferSize - parser.BufferOffset;
+                    var tmpBuffer = new byte[bufferSize];
+                    Array.Copy(buffer.Buffer.Array ?? throw new InvalidOperationException(), buffer.Buffer.Offset + parser.BufferOffset, tmpBuffer, 0, bufferSize);
+
+                    streamInfo = new AttachmentsStreamInfo
+                    {
+                        AttachmentAdvancedDetails = attachmentsMetadata,
+                        Buffer = tmpBuffer
+                    };
+                }
+
+                Result = attachmentsMetadata.AttachmentsMetadata.Select(attachment => new AttachmentEnumeratorResult(new AttachmentsStream(stream, attachment.Name, attachment.Index, attachment.Size, streamInfo))
```

Comment:
You can just return an enumerator of `LimitedStream` here, and that would serve the exact same purpose.
The only complexity is the fact that you are using the buffer, and that can be handled by concat stream more easily

Reply:
Or at least maybe we can inherit from `LimitedStream` (was not aware that we have sth like that) if you need to to additional work?

Reply:
`LimitedStream` adjust and uses Position, we don't have position in the inner stream(gzip) of attachmentsStream.